### PR TITLE
Node http api improvements

### PR DESCRIPTION
**Pull request by nickeskov**

### CODE REVIEW

File: pkg/api/app.go [hunk 1]
```diff
@@ -97,28 +97,32 @@ func (a *App) LoadKeys(apiKey string, password []byte) error {
...
 func (a *App) checkAuth(key string) error {
+	// TODO(nickeskov): use new types of errors
 	if !a.apiKeyEnabled {
 		return &AuthError{errors.New("api key disabled")}
 	}
 	d, err := crypto.SecureHash([]byte(key))
 	if err != nil {
-		return &AuthError{err}
+		// TODO(nickeskov): it's OK?
+		return err
```
Comment (esuwu) [inline pkg/api/app.go:125]: you wanted to use custom errors in order to make switch on them. Why is here just a regular error?

File: pkg/api/app_miner.go [hunk 2]
```diff
@@ -19,18 +19,18 @@ type MinerInfo struct {
 	Scheduler Scheduler
 }
 
-func (a *App) Miner() (*MinerInfo, error) {
+func (a *App) Miner() (MinerInfo, error) {
```
Comment (esuwu) [inline pkg/api/app_miner.go:22]: this function is not going to return an error. There is always `nil`. Should be `func (a *App) Miner() MinerInfo`

File: pkg/api/app_peers.go [hunk 3]
```diff
@@ -14,22 +14,22 @@ type Peer struct {
...
+type PeersKnown struct {
 	Peers []Peer `json:"peers"`
 }
 
-func (a *App) PeersAll() (*PeersAll, error) {
+func (a *App) PeersKnown() (PeersKnown, error) {
 	peers, err := a.state.Peers()
 	if err != nil {
-		return nil, &InternalError{err}
+		return PeersKnown{}, errors.Wrap(err, "PeersKnown")
```
Comment (esuwu) [inline pkg/api/app_peers.go:24]: the returned error should be errors.Wrap(err, "failed to get known peers")

File: pkg/api/app_peers.go [hunk 4]
```diff
@@ -73,24 +73,29 @@ type PeersConnectedRow struct {
 	ApplicationVersion string `json:"applicationVersion"`
 }
 
-func (a *App) PeersConnected() (*PeersConnectedResponse, error) {
+func (a *App) PeersConnected() (PeersConnectedResponse, error) {
```
Comment (esuwu) [inline pkg/api/app_peers.go:76]: this function is also not going to return an error

File: pkg/api/app_peers.go [hunk 5]
```diff
@@ -99,16 +104,16 @@ type PeersSuspendedResponse struct {
 	Peers []string `json:"peers"`
 }
 
-func (a *App) PeersSuspended() (*PeersSuspendedResponse, error) {
+func (a *App) PeersSuspended() (PeersSuspendedResponse, error) {
```
Comment (esuwu) [inline pkg/api/app_peers.go:107]: the same about errors

Comment (esuwu) [reply pkg/api/app_peers.go:107]: are you sure there should be a particular function for calling `a.peers.Suspended()`?

File: pkg/api/errors.go [hunk 6]
```diff
@@ -11,3 +21,69 @@ type AuthError struct {
...
+func (eh *ErrorHandler) Handle(w http.ResponseWriter, r *http.Request, err error) {
+	if err == nil {
+		return
+	}
+	switch innerErr := errors.Cause(err).(type) {
+	case BadRequestError, *BadRequestError:
+		// TODO(nickeskov): remove this
+		http.Error(w, fmt.Sprintf("Failed to complete request: %s", innerErr.Error()), http.StatusForbidden)
+	case AuthError, *AuthError:
+		// TODO(nickeskov): remove this
```
Comment (esuwu) [inline pkg/api/errors.go:44]: `// TODO(nickeskov): remove this` when exactly are you going to remove `this`? delete comments here or remove something that you was going to

File: pkg/api/errors.go [hunk 7]
```diff
@@ -11,3 +21,69 @@ type AuthError struct {
...
+	w.WriteHeader(apiErr.GetHttpCode())
+	if encodeErr := json.NewEncoder(w).Encode(apiErr); encodeErr != nil {
+		eh.logger.Error("Failed to marshal API Error to JSON",
+			zap.String("proto", r.Proto),
+			zap.String("path", r.URL.Path),
+			zap.String("request_id", middleware.GetReqID(r.Context())),
+			zap.Error(encodeErr),
+			zap.String("api_error", apiErr.Error()),
+		)
+		panic(errors.Errorf("BUG, CREATE REPORT: %s", encodeErr.Error()))
```
Comment (esuwu) [inline pkg/api/errors.go:87]: should we really panic here?

File: pkg/api/errors/transaction.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import (
+	"github.com/wavesplatform/gowaves/pkg/proto"
+	"net/http"
+)
+
+type transactionError struct {
+	genericError
+}
+
+// TODO(nickeskov): IMPLEMENT ME
```
Comment (esuwu) [inline pkg/api/errors/transaction.go:12]: should it be implemented if it's not used?

File: pkg/api/app_blocks.go [hunk 9]
```diff
@@ -9,12 +10,12 @@ type Score struct {
 	Score string `json:"score"`
 }
 
-func (a *App) BlocksScoreAt(at proto.Height) (*Score, error) {
+func (a *App) BlocksScoreAt(at proto.Height) (Score, error) {
```
Comment (esuwu) [inline pkg/api/app_blocks.go:13]: I don't think here should be a function for just `score, err := a.state.ScoreAtHeight(at)`

File: pkg/api/errors.go (hunk 7 above)
Comment (nickeskov) [reply pkg/api/errors.go:87]: Yes, we should. Type which implements `ApiError` interface must be serializable to JSON. If encode fails that is definitely a bug.

File: pkg/api/app.go (hunk 1 above)
Comment (nickeskov) [reply pkg/api/app.go:125]: Regular errors will be automatically transformed into custom unknown error, so here I want to use regular error.

File: pkg/api/app_blocks.go (hunk 9 above)
Comment (nickeskov) [reply pkg/api/app_blocks.go:13]: I don't think so because this layer hides node and state internals.

File: pkg/api/errors/transaction.go (hunk 8 above)
Comment (nickeskov) [reply pkg/api/errors/transaction.go:12]: Yes because we need to support scala node HTTP API.

File: pkg/api/app_peers.go (hunk 3 above)
Comment (alexeykiselev) [reply pkg/api/app_peers.go:55]: Better to change error message here

**Review by alexeykiselev — CHANGES_REQUESTED** [review]
Please, rebase on `version-0.9` branch

**Comment by nickeskov** [comment]
See #475

**Pull request closed without merging**